### PR TITLE
Add ecmult_gen, ecmult_const and ecmult to benchmark

### PR DESCRIPTION
**Pull request by jonasnick**
I was trying to determine the impact of ecmult_gen in schnorrsig signing and noticed that there is no way to bench this right now. The new benchmarks look like this:
```
$ ./bench_ecmult
ecmult_gen: min 20.9us / avg 21.2us / max 21.7us
ecmult_const: min 63.9us / avg 64.3us / max 64.8us
ecmult 1: min 49.4us / avg 49.7us / max 50.3us
ecmult 1g: min 39.8us / avg 40.0us / max 40.3us
ecmult 2g: min 27.2us / avg 27.3us / max 27.8us
ecmult_multi 1g: min 39.8us / avg 40.0us / max 40.2us
ecmult_multi 2g: min 27.2us / avg 27.4us / max 27.7us
ecmult_multi 3g: min 22.8us / avg 22.9us / max 23.1us
ecmult_multi 4g: min 20.6us / avg 20.8us / max 21.1us
ecmult_multi 5g: min 19.3us / avg 19.5us / max 19.7us
```

(Turns out ecmult_gen is 37% of the 55.8us that schnorrsig sign takes)

### CODE REVIEW

File: src/bench_ecmult.c [hunk 1]
```diff
@@ -34,15 +35,144 @@ typedef struct {
...
 
     /* Test output. */
     secp256k1_gej* output;
 } bench_data;
 
-static int bench_callback(secp256k1_scalar* sc, secp256k1_ge* ge, size_t idx, void* arg) {
+/* Hashes x into [0, POINTS) twice and store the result in offset1 and offset2. */
+static void hash_into_offset(bench_data* data, size_t x) {
+    data->offset1 = (x * 0x537b7f6f + 0x8f66a481) % POINTS;
+    data->offset2 = (x * 0x7f6f537b + 0x6a1a8f49) % POINTS;
```
Comment (elichai) [inline src/bench_ecmult.c]: where's this thing from?
I see it's also in `bench_ecmult_setup` but can't find the source of this

Comment (jonasnick) [reply src/bench_ecmult.c]: I think is this is just a standard integer hash function (akin to https://en.wikipedia.org/wiki/Universal_hashing#Hashing_integers) whose parameters sipa selected. But looking at it now it may be give poor results because `POINTS` is not prime.

Comment (gmaxwell) [reply src/bench_ecmult.c]: I don't think points needs to be prime, so long as the multipliers are coprime to it.

Comment (jonasnick) [reply src/bench_ecmult.c]: fwiw they're all coprime to POINTS.

File: src/bench_ecmult.c [hunk 2]
```diff
@@ -34,15 +35,144 @@ typedef struct {
...
+        secp256k1_ecmult(&data->ctx->ecmult_ctx, &data->output[iter], &data->pubkeys_gej[(data->offset1+iter) % POINTS], &data->scalars[(data->offset2+iter) % POINTS], NULL);
+    }
+}
+
+static void bench_ecmult_1_teardown(void* arg) {
+    bench_data* data = (bench_data*)arg;
+    bench_ecmult_teardown_helper(data, &data->offset1, &data->offset2, NULL);
+}
+
+static void bench_ecmult_1g(void* arg) {
```
Comment (real-or-random) [inline src/bench_ecmult.c]:
```suggestion
static void bench_ecmult_0g(void* arg) {
```

File: src/bench_ecmult.c [hunk 3]
```diff
@@ -34,15 +35,144 @@ typedef struct {
...
+    secp256k1_scalar zero;
+    size_t iter;
+
+    secp256k1_scalar_set_int(&zero, 0);
+    for (iter = 0; iter < ITERS; ++iter) {
+        secp256k1_ecmult(&data->ctx->ecmult_ctx, &data->output[iter], NULL, &zero, &data->scalars[(data->offset1+iter) % POINTS]);
+    }
+}
+
+static void bench_ecmult_1g_teardown(void* arg) {
```
Comment (real-or-random) [inline src/bench_ecmult.c]:
```suggestion
static void bench_ecmult_0g_teardown(void* arg) {
```

and so on..

File: src/bench_ecmult.c [hunk 4]
```diff
@@ -34,15 +35,144 @@ typedef struct {
...
+    bench_data* data = (bench_data*)arg;
+    bench_ecmult_teardown_helper(data, NULL, NULL, &data->offset1);
+}
+
+static void bench_ecmult_2g(void* arg) {
```
Comment (real-or-random) [inline src/bench_ecmult.c]: 1g

**Comment by jonasnick** [comment]
@real-or-random I don't think so (unless we change it for ecmult_multi) but it took me a while to figure this out again. I added a commit that printfs and explanation. It's not pretty, but less confusing.

**Comment by jonasnick** [comment]
squashed

File: src/bench_ecmult.c [hunk 5]
```diff
@@ -129,21 +241,37 @@ static void run_test(bench_data* data, size_t count, int includes_g) {
...
     /* Run the benchmark. */
-    sprintf(str, includes_g ? "ecmult_%ig" : "ecmult_%i", (int)count);
-    run_benchmark(str, bench_ecmult, bench_ecmult_setup, bench_ecmult_teardown, data, 10, count * (1 + ITERS / count));
+    sprintf(str, includes_g ? "ecmult_multi %ig" : "ecmult_multi %i", (int)count);
+    run_benchmark(str, bench_ecmult_multi, bench_ecmult_multi_setup, bench_ecmult_multi_teardown, data, 10, count * (1 + ITERS / count));
+}
+
+static void generate_scalar(uint32_t num, secp256k1_scalar* scalar) {
+    secp256k1_sha256 sha256;
+    unsigned char c[11] = {'e', 'c', 'm', 'u', 'l', 't', 0, 0, 0, 0};
```
Comment (real-or-random) [inline src/bench_ecmult.c]: I know it's just code you move around but this annoys me every time I see it...
```suggestion
    unsigned char c[10] = {'e', 'c', 'm', 'u', 'l', 't', 0, 0, 0, 0};
```

It's not a bug because `c[9]` will be initialized to 0 automatically. But it's a code smell and it can't be wrong to fix it. 

ACK for the rest.

Comment (jonasnick) [reply src/bench_ecmult.c]: Added commit with fix

**Comment by real-or-random** [comment]
ACK a419f3dddfc96b0c7454985a913b932b77b94b0b  I read the code and tried it

File: src/bench_ecmult.c [hunk 6]
```diff
@@ -128,65 +265,71 @@ static void run_test(bench_data* data, size_t count, int includes_g, int num_ite
...
         } else if(have_flag(argc, argv, "simple")) {
-            printf("Using simple algorithm:\n");
+            printf("Using simple combined algorithm:\n");
             data.ecmult_multi = secp256k1_ecmult_multi_var;
-            secp256k1_scratch_space_destroy(data.ctx, data.scratch);
-            data.scratch = NULL;
         } else {
             fprintf(stderr, "%s: unrecognized argument '%s'.\n", argv[0], argv[1]);
-            fprintf(stderr, "Use 'pippenger_wnaf', 'strauss_wnaf', 'simple' or no argument to benchmark a combined algorithm.\n");
+            fprintf(stderr, "Use '--help', 'pippenger_wnaf', 'strauss_wnaf', 'simple' or no argument to benchmark a combined algorithm.\n");
```
Comment (real-or-random) [inline src/bench_ecmult.c:297]: simply call `help` here? Also, this is slightly confusing because it does not make clear that `simple` and no argument are the same thing.

File: src/bench_ecmult.c [hunk 7]
```diff
@@ -128,65 +265,71 @@ static void run_test(bench_data* data, size_t count, int includes_g, int num_ite
...
             data.ecmult_multi = secp256k1_ecmult_pippenger_batch_single;
         } else if(have_flag(argc, argv, "strauss_wnaf")) {
             printf("Using strauss_wnaf:\n");
             data.ecmult_multi = secp256k1_ecmult_strauss_batch_single;
         } else if(have_flag(argc, argv, "simple")) {
-            printf("Using simple algorithm:\n");
+            printf("Using simple combined algorithm:\n");
             data.ecmult_multi = secp256k1_ecmult_multi_var;
-            secp256k1_scratch_space_destroy(data.ctx, data.scratch);
-            data.scratch = NULL;
```
Comment (real-or-random) [inline src/bench_ecmult.c:159]: That line got lost, and this was what made the difference between simple and combined? I see you merged this into "simple combined" in the printfs. Was this change on purpose?

Comment (jonasnick) [reply src/bench_ecmult.c:159]: Good catch. I also forgot what the simple algorithm actually does. I fixed this and added an explanation to the help().

File: src/bench_ecmult.c (hunk 6 above)
Comment (jonasnick) [reply src/bench_ecmult.c]: okay

**Comment by jonasnick** [comment]
@real-or-random care to  reACK :) ?

File: src/bench_ecmult.c (hunk 6 above)
Comment (real-or-random) [reply src/bench_ecmult.c]: Ah sorry, my suggestion was just to replace the entire `else` branch with `help()`, not to replace the `--help` by `help`. Feel free to change back, or to ignore.

**Comment by real-or-random** [comment]
ACK ee83dbf267b34bb4a01c872cd02bf1c8065e3166

File: src/bench_ecmult.c (hunk 6 above)
Comment (jonasnick) [reply src/bench_ecmult.c]: Fixed

**Comment by jonasnick** [comment]
Rebased to let CI do its thing

**Comment by sipa** [comment]
While we're at it: are no-g variants not interesting?

**Comment by jonasnick** [comment]
Hm with-g seems more interesting (schnorrsig batch verify, taproot tweaks, pedersen commitments,...) but we could add a flag or something for no-g (preferably in a separate PR imho).

**Comment by real-or-random** [comment]
ACK 8f879c2887e166da2ec959ce78078f7b84ebfdf9

**Comment by elichai** [comment]
tACK 8f879c2887e166da2ec959ce78078f7b84ebfdf9

nit, It is a little confusing that the first 4 benchmarks are unrelated to the flag being passed (Maybe we want to move the "Using XXX algorithm" print to just before the relevant benchmarks?)

**Comment by real-or-random** [comment]
> nit, It is a little confusing that the first 5 benchmarks are unrelated to the flag being passed (Maybe we want to move the "Using XXX algorithm" print to just before the relevant benchmarks?)

Indeed but this can be done in a separate PR, too.

Let me take the opportunity to merge this PR now that we have two ACKs.
